contracts/core.clar

```
@@ -0,0 +1,180 @@
+    votes: uint
+  }
+)
+
+(define-map MiningCandidateVoters
+  { contract: principal, voter: principal }
+  bool
+)
+
+(define-read-only (has-voted-on-candidate (contract principal))
+  (is-some (map-get? MiningCandidateVoters
+    { contract: contract, voter: contract-caller }
+  ))
+)
+
+(define-read-only (get-city-wallet)
+  (var-get cityWallet)
+)
+
+(define-public (set-city-wallet (newCityWallet principal))
+  (begin
+    (asserts! (is-authorized) (err ERR_UNAUTHORIZED))
+    (ok (var-set cityWallet newCityWallet))
+  )
+)
+
+(define-read-only (get-mining-contract (address principal))
+  (map-get? MiningContracts address)
+)
+
+(define-public (add-mining-contract (address principal))
+  (let
+    (
+      (newNonce (+ (var-get miningContractNonce) u1))
+    )
+    (asserts! (is-authorized) (err ERR_UNAUTHORIZED))
+    (asserts! (not (is-eq (get-active-mining-contract) address)) (err ERR_CONTRACT_ALREADY_EXISTS))
+    (asserts! (is-none (get-mining-contract address)) (err ERR_CONTRACT_ALREADY_EXISTS))
+    
+    (var-set miningContractNonce newNonce)
+    (map-set MiningContractVotes 
+      newNonce
+      {
+        address: address,
+        startBH: (+ block-height u1),
+        endBH: (+ block-height u1 DEFAULT_VOTING_PERIOD),
+        miners: u0,
+        votes: u0
+      }
+    )
+    (ok (map-set MiningContracts address { id: newNonce, state: STATE_DEFINED } ))
+  )
+)
+
+(define-read-only (get-mining-contract-vote (contractId uint))
+  (map-get? MiningContractVotes contractId)
+)
+
+(define-public (vote-on-mining-contract (contract principal))
+  (let
+    (
+      (contractId (get id (unwrap! (get-mining-contract contract) (err ERR_CONTRACT_DO_NOT_EXISTS))))
+      (contractVote (unwrap-panic (get-mining-contract-vote contractId)))
+    )
+    (asserts! (is-between block-height (get startBH contractVote) (get endBH contractVote)) 
+      (err ERR_VOTE_HAS_ENDED))
+    (asserts! (not (has-voted-on-candidate contract))
+      (ok false))
```

Why return `ok false` here over throwing an error like `ERR_ALREADY_VOTED` or similar?

---

Great finding.
At first I wanted to suppress all errors in this function as they would fail `mine-tokens` that will be responsible for calling it, but later I changed my mind and forgot to fix this execution path.
I'll add proper error with next commit.

---

Ah... that's interesting! I hadn't thought of it through the path of calling via `mine-tokens` and instead was looking at it as an independent function.

I think this still makes sense to error out though since it follows the same pattern as all of our other checks and balances. The UI can be updated to check `has-voted-on-candidate` before displaying a vote button or similar.

---

Fiexed with https://github.com/citycoins/citycoin/pull/111/commits/0222f6310e5c9745dbc4360482ae21294be04035